licensing/src/main/java/io/crate/license/TransportSetLicenseAction.java

```diff
@@ -0,0 +1,83 @@
+ * obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+ * implied.  See the License for the specific language governing
+ * permissions and limitations under the License.
+ *
+ * However, if you have executed another commercial license agreement
+ * with Crate these terms will supersede the license and you may use the
+ * software solely pursuant to the terms of the relevant commercial
+ * agreement.
+ */
+
+package io.crate.license;
+
+import org.elasticsearch.action.ActionListener;
+import org.elasticsearch.action.support.ActionFilters;
+import org.elasticsearch.action.support.master.TransportMasterNodeAction;
+import org.elasticsearch.cluster.ClusterState;
+import org.elasticsearch.cluster.block.ClusterBlockException;
+import org.elasticsearch.cluster.block.ClusterBlockLevel;
+import org.elasticsearch.cluster.metadata.IndexNameExpressionResolver;
+import org.elasticsearch.cluster.service.ClusterService;
+import org.elasticsearch.common.inject.Inject;
+import org.elasticsearch.common.inject.Singleton;
+import org.elasticsearch.common.settings.Settings;
+import org.elasticsearch.threadpool.ThreadPool;
+import org.elasticsearch.transport.TransportService;
+
+@Singleton
+public class TransportSetLicenseAction
+    extends TransportMasterNodeAction<SetLicenseRequest, SetLicenseResponse> {
+
+    private final LicenseService licenseService;
+
+    @Inject
+    public TransportSetLicenseAction(Settings settings,
+                                     TransportService transportService,
+                                     ClusterService clusterService,
+                                     ThreadPool threadPool,
+                                     LicenseService licenseService,
+                                     ActionFilters actionFilters,
+                                     IndexNameExpressionResolver indexNameExpressionResolver) {
+        super(settings, "crate/sql/set_license", transportService, clusterService, threadPool, actionFilters,
+            indexNameExpressionResolver, SetLicenseRequest::new);
+        this.licenseService = licenseService;
+    }
+
+    @Override
+    protected String executor() {
+        return ThreadPool.Names.GENERIC;
+    }
+
+    @Override
+    protected SetLicenseResponse newResponse() {
+        return new SetLicenseResponse();
+    }
+
+    @Override
+    protected void masterOperation(final SetLicenseRequest request,
+                                   ClusterState state,
+                                   ActionListener<SetLicenseResponse> listener) throws Exception {
+
+        LicenseMetaData metaData = request.licenseMetaData();
+        if (licenseService.validateLicense(metaData)) {
```

Comment:
We probably have to take care that the `listener` is also invoked if the validation fails, otherwise this might result in stuck queries.